api/proto/auth/grpc.go

```diff
@@ -0,0 +1,932 @@
+		cancel()
+		return nil, trail.FromGRPC(err)
+	}
+	k := &streamKeepAliver{
+		stream:      stream,
+		ctx:         cancelCtx,
+		cancel:      cancel,
+		keepAlivesC: make(chan services.KeepAlive),
+	}
+	go k.forwardKeepAlives()
+	go k.recv()
+	return k, nil
+}
+
+type streamKeepAliver struct {
+	sync.RWMutex
+	stream      AuthService_SendKeepAlivesClient
+	ctx         context.Context
+	cancel      context.CancelFunc
+	keepAlivesC chan services.KeepAlive
+	err         error
+}
+
+func (k *streamKeepAliver) KeepAlives() chan<- services.KeepAlive {
+	return k.keepAlivesC
+}
+
+func (k *streamKeepAliver) forwardKeepAlives() {
+	for {
+		select {
+		case <-k.ctx.Done():
+			return
+		case keepAlive := <-k.keepAlivesC:
+			err := k.stream.Send(&keepAlive)
+			if err != nil {
+				k.closeWithError(trail.FromGRPC(err))
+				return
+			}
+		}
+	}
+}
+
+func (k *streamKeepAliver) Error() error {
+	k.RLock()
+	defer k.RUnlock()
+	return k.err
+}
+
+func (k *streamKeepAliver) Done() <-chan struct{} {
+	return k.ctx.Done()
+}
+
+// recv is necessary to receive errors from the
+// server, otherwise no errors will be propagated
+func (k *streamKeepAliver) recv() {
+	err := k.stream.RecvMsg(&empty.Empty{})
+	k.closeWithError(trail.FromGRPC(err))
+}
+
+func (k *streamKeepAliver) closeWithError(err error) {
+	k.Close()
+	k.Lock()
+	defer k.Unlock()
+	k.err = err
+}
+
+func (k *streamKeepAliver) Close() error {
+	k.cancel()
+	return nil
+}
+
+// NewWatcher returns a new event watcher
+func (c *Client) NewWatcher(ctx context.Context, watch services.Watch) (services.Watcher, error) {
+	if err := c.ConnectGRPC(); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	cancelCtx, cancel := context.WithCancel(ctx)
+	var protoWatch Watch
+	for _, k := range watch.Kinds {
+		protoWatch.Kinds = append(protoWatch.Kinds, WatchKind{
+			Name:        k.Name,
+			Kind:        k.Kind,
+			LoadSecrets: k.LoadSecrets,
+			Filter:      k.Filter,
+		})
+	}
+	stream, err := c.grpc.WatchEvents(cancelCtx, &protoWatch)
+	if err != nil {
+		cancel()
+		return nil, trail.FromGRPC(err)
+	}
+	w := &streamWatcher{
+		stream:  stream,
+		ctx:     cancelCtx,
+		cancel:  cancel,
+		eventsC: make(chan services.Event),
+	}
+	go w.receiveEvents()
+	return w, nil
+}
+
+type streamWatcher struct {
+	sync.RWMutex
+	stream  AuthService_WatchEventsClient
+	ctx     context.Context
+	cancel  context.CancelFunc
+	eventsC chan services.Event
+	err     error
+}
+
+func (w *streamWatcher) Error() error {
+	w.RLock()
+	defer w.RUnlock()
+	if w.err == nil {
+		return trace.Wrap(w.ctx.Err())
+	}
+	return w.err
+}
+
+func (w *streamWatcher) closeWithError(err error) {
+	w.Close()
+	w.Lock()
+	defer w.Unlock()
+	w.err = err
+}
+
+func (w *streamWatcher) Events() <-chan services.Event {
+	return w.eventsC
+}
+
+func (w *streamWatcher) receiveEvents() {
+	for {
+		event, err := w.stream.Recv()
+		if err != nil {
+			w.closeWithError(trail.FromGRPC(err))
+			return
+		}
+		out, err := eventFromGRPC(*event)
```

Comment:
We should move `eventFromGRPC` or maybe `streamWatcher` out, that removes the other dependency on the `events` package.

Reply:
The `events` package is only being used for `AuditStreams`, but I'll try to factor out events in another pr.